### PR TITLE
Resolves some Battleground issues

### PR DESCRIPTION
**Pull request by aleos89**
* **Addressed Issue(s)**: #4665

* **Server Mode**: Pre-renewal and Renewal

* **Description of Pull Request**: 
  * Queue data will now be kept available until a Battleground event is over.
  * Players can now join an active Battleground.
  * When a player logs off or leaves a queue they will now properly be removed from the queue.
  * Fixes a typo in Flavius 2 during unbooking process.
  * Better documents script commands bg_reserve and bg_unbook.
  * Other small fixes and cleanups.
Thanks to @roSBK and @Daegaladh!

### CODE REVIEW

**Comment by roSBK** [comment]
For the tests I did ok, but in the desert where is saved the time of punishment for the new record? Also a question of Max Player in the db, I don't know if it was to work in an expansive way, but it is not functional, and if it is not to work in an expansive way it is completely unnecessary.

**Comment by reunite-ro** [comment]
Hello, is it normal that we need to relog after the KVM round ? Will popup duplicate application for party leader or nothing popup for party member.

**Comment by roSBK** [comment]
Crash in line: pc_setpos(sd, bgteam->cemetery.map, bgteam->cemetery.x, bgteam->cemetery.y, CLR_TELEPORT);

**Comment by roSBK** [comment]
So I applied everything else or the error continues:

https://i.imgur.com/R6xOXeP.png

**Comment by admkakaroto** [comment]
crash on map_server using this PR, if playing in 'party' mode.

**Comment by aleos89** [comment]
What steps are you doing to create this crash point? Can you provide me with some more details.

**Comment by admkakaroto** [comment]
after the patch, the bg modes do not start even though there are more players than necessary in the queue:
a1

**Comment by admkakaroto** [comment]
We were able to play a game in 'individual' mode, but after finishing we were unable to start another game, it displays the message 'Duplicate Registration' whenever we try to register again.

dupl

On map_server:
```
[Error]: bg_queue_leave: Couldn't find player Ninguem in battlegrounds queue.
[Error]: bg_queue_leave: Couldn't find player Momo in battlegrounds queue.
```

**Comment by roSBK** [comment]
I put the code inspection on the imgur line by line^^

**Comment by admkakaroto** [comment]
Little warning:
`\rAthena\src\map\battleground.cpp(1001,14): warning C4244: 'initializing': conversion from 'int64' to 'int', possible loss of data`

Code:

> 		// Enough players have joined
> 		if (queue->map && queue->map->isReserved) { // Battleground is already active
> 			int bg_id = mapreg_readreg(add_str(r ? queue->map->team2.bg_id_var.c_str() : queue->map->team1.bg_id_var.c_str()));


Print:
a1

**Comment by admkakaroto** [comment]
Issue:
```
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
[Error]: clif_bg_queue_apply_notify: Player is not in a battleground queue.
```


How to reproduce:

- Start 'individual' battleground match.
- After finish the match, try start a new.

**Comment by admkakaroto** [comment]
The problem is that a player can enter the BG after it is finished, he is then stuck on the map without being able to leave, receives no reward and is not sent out when talking to the npc.

a1


To reproduce:

- Start KVM.
- When the game is over, try to register another character in the queue, but it must be a character who has not played the previous game.

it will be recaled to the arena where the previous players were, but he is unable to leave or receive a reward.

After 1 minute he is kicked out.

But he can no longer open the BG window and no way starts the battle anymore.

**Comment by admkakaroto** [comment]
Well now some things have worked but others still haven't, after finishing a BG mode and trying to start the same mode or another game mode again, it just doesn't start, even with the right number of players or more than necessary, follow print :

a11

**Comment by roSBK** [comment]
Well, now it seems that the expansive mode has worked, for the most part, that it is unbalanced, come on sooner, I have to see if you are in pairs, and if you are not teleporting the correct map, you sent me for or map queue_qsch05.

I am sorry that you should add more configuration for new player spawn.

**Comment by roSBK** [comment]
Well, there was another bug too, in the first record he doesn't go to the arena, then what happens, he has to go out with the characters in the queue, remove the sc_data timer and then register again. After it is over and the player is going to register, he says that the order was a duplicate which makes him have to log out and come back again, even so he still had an error saying he should try again soon afterwards he added to the queue and again the previous bug, you have to remove the player and the timer ...

Here are the error messages on the emulator console:
[Warning]: buildin_bg_unbook: Could not unreserve battleground with name bat_c01

[Error]: bg_queue_leave: Couldn't find player DestyX in battlegrounds queue.

**Comment by aleos89** [comment]
The duplicate registration is probably because the previous battleground is still active (in the period where players are getting their prizes).

@admkakaroto I was able to join again after this period ended (individual and party). Make sure you have all the recent changes from this PR.

**Comment by admkakaroto** [comment]
Well, when the characters login into the game and try to enter the BG, some need to wait 1 minute to register, others can enter without wait.

We tried the first time with 12 characters, it didn't start:
a1111

I turned off the emulator, turned it on again and this time started the first time, but after leaving it simply did not start Bg a new battle:
a2222

**Comment by admkakaroto** [comment]
Well, I tested KVM about 3 times and it always started normally, but it was necessary to put 11 players and not 10, as requested.

**Comment by aleos89** [comment]
Make sure your logic checks for required players are good. My local tests started started off with the minimum required. I lowered the count to 2v2 but it still kicked off with 4 players.

**Comment by admkakaroto** [comment]
Okay, I'm going to test again, changing to 2x2 and see if it works.

**Comment by admkakaroto** [comment]
`[Error]: delete_timer error : function mismatch 01583D30(connect_check_clear) != 013A8B30(bg_on_ready_active)`

**Comment by Daegaladh** [comment]
A couple of bugs more:
When all the players are ready and they're teleported to the BG this error shows up
`[ Error ] : delete_timer error : function mismatch 0x5559366b0650(connect_check_clear) != 0x55593647e9e0(bg_on_ready_active)`
same as @admkakaroto

If 2 or more players left the BG, and another player tries to join, the message that the BG is about to start pops-up but isn't teleported.

**Comment by Daegaladh** [comment]
Previus bugs are fixed, but the teams still aren't balanced. I've tested 2vs2, and the 2 from one team left, and 2 new players joined, one ended up in Croix and the other in Guillaume, so it became 3vs1

**Comment by Artuvazro** [comment]
Bugs detected in the latest version:

· Flavius individual:
1.- 6 players.
2.- 4 enter in teams of 2.
3.- 2 people leave.
4.- The other 2 join.
5.- The first round is player properly. A crystal is destroyed.
6.- Second round starts... **MAYHEM**
 Teams are destroyed serverside. All players can attack all players regardless of their team. Guardians attack every player including the ones from your team. Every player can destroy every crystal. The script crashes when a crystal is destroyed and the battleground flow is interrupted. No error is displayed in the console.

File: npc/battleground/flavius/flavius01.txt [hunk 1]
```diff
@@ -225,6 +225,8 @@ OnMyMobDead:
 			enablenpc "Guillaume Vintenar#b01_a";
 			enablenpc "Croix Vintenar#b01_b";
 			donpcevent "time#bat_b01::OnStop";
+			if (getbattleflag("feature.bgqueue"))
+				bg_reserve "bat_b01", true;
```
Comment (Daegaladh) [inline npc/battleground/flavius/flavius01.txt:229]: You forgot the other crystal in both files

**Comment by Artuvazro** [comment]
A new bug:

If a player clicks on Battle Therapist and leaves the window open, that player is not warped to the battleground. That could be ok, it's the player fault. But... problem is that if the player closes the NPC window he's not warped to the battlefield in any of the posterior warps and becomes trapped. This is fixed if you recall that player to the battlefield and kill him again.

**Comment by Artuvazro** [comment]
Another thing:

I suspect that teams are not randomized. We played three games (flavius individual) and all of us ended in the same teams paired with the same people in the three games. Is a variable not being cleaned or something like that?

**Comment by Artuvazro** [comment]
A further bug (please, don't hate me):

If you don't press OK when asked to join the BG (once teams are ready and you're about to get warped to the battlefield), the player is never allowed to join another BG. He gets a message stating that he must logout and log in again but this doesn't solve the issue either. The only solution is to restart the server.

**Comment by Artuvazro** [comment]
One more (overkill):

/battlechat is bugged. When enabled, one member of the team can read and write in the chat, while the other team member can only write but not read himself or the the other player text. And additionally, both players send their text duplicated to the chat.

**Comment by aleos89** [comment]
> Another thing:
> 
> I suspect that teams are not randomized. We played three games (flavius individual) and all of us ended in the same teams paired with the same people in the three games. Is a variable not being cleaned or something like that?

```cpp
bool r = rnd() % 2 != 0;
```
It's a 50/50 chance on what team you're put on. When a BG queue is destroyed (through timeout or normal ending) the team data is completely destroyed.

**Comment by Artuvazro** [comment]
> 
> 
> A further bug (please, don't hate me):
> 
> If you don't press OK when asked to join the BG (once teams are ready and you're about to get warped to the battlefield), the player is never allowed to join another BG. He gets a message stating that he must logout and log in again but this doesn't solve the issue either. The only solution is to restart the server.

This was almost fixed, but not completely.
When a player doesn't press OK, he gets a cool down of 1 minute. That's good, and this was fixed because previously you were stuck. The thing is that now, all the other players that pressed OK are stuck and can't join any other battleground. The button from the UI doesn't do anything until you logout.

**Comment by Artuvazro** [comment]
A new bug:
I joined a battleground (flavius individual) that had already started. When this happened, the team which I joined was warped to their respawn area. The other team didn't. Is this ok? It's a bit exploitable, as you can join a battleground and spoil the game of your team by warping them to their respawn for no apparent reason.

**Comment by Artuvazro** [comment]
Flavius individual. Teams are destroyed again.

How to trigger:
Four people join the game in 2vs2. A fifth person joins. A crystal is broken in the first round and the round finishes normally. In the next round, all guardians attack everybody and can be killed by everybody. The BG flown is broken. Teams are probably destroyed.

**Comment by Artuvazro** [comment]
Hi @aleos89 everything is fixed, good job. Except for the broken teams issue. However, we found how to reproduce it:

**2vs2**
**- Round 1:**
· _Soome one else joins the BG_ (3vs2).
· A team breaks the Crystal.

**- Round 2:**
· Teams are destroyed.

We tried and it can be reproduced in Round 3 if in Round 2 someone joined. Therefore, the rule is that it gets broken in the round next to someone joining a BG that had already started.

**Comment by Artuvazro** [comment]
A new bug:

We found that there are some warps that don't make sense. When a crystal is broken, teams are warped to the initial cell in the map and then instantly warped to the respawn area.

Lines 147 and 148 in the script. But I think that the whole OnReset: label should be revised...

**Comment by aleos89** [comment]
The teams getting destroyed is strange. I seemed to have gotten that case once but then I can't trigger it again, even following your method. I'll keep trying different scenarios.

**Comment by roSBK** [comment]
Have these codes been applied to the current rathena project? it would be interesting, because adding everything that was modified there for me to test again is kind of complicated, and the one that is there is no longer working.

**Comment by aleos89** [comment]
There’s one more issue with a timer for Flavius that needs to be fixed.

**Pull request merged**